### ISSUE TITLE
Gradle daemon locks files by leaving file handles open

### ISSUE THREAD

**Issue by veita** [gradle/gradle#937]
It happens that gradle daemon leaves file handles open after a build has finished. This prevents some files to be accessed or deleted.

### Expected Behavior
Gradle daemon should not lock files.

### Current Behavior
Gradle daemon locks some files. This can be verified by identifying the locking process with Sysinternals Process Explorer.

### Context
Tried to build the Groovy project using `gradle build`.

### Steps to Reproduce (for bugs)
1. `git clone git://git.apache.org/groovy.git`
2. `cd groovy`
3. Check out commit 273c15375bb8ab03e9ac17a13818bef69491dc9c if necessary.
4. `gradle build` (the build fails).
5. `git clean -d -x -f`

The last command fails since there are locked files. Use Process Explorer to find out the which process locks the files. The offending process is
```
java.exe -XX:MaxPermSize=384m -XX:+CMSClassUnloadingEnabled -XX:+CMSPermGenSweepingEnabled -Xmx1G -Dfile.encoding=windows-1252 -Duser.country=DE -Duser.language=de -Duser.variant -cp C:\java\gradle\lib\gradle-launcher-3.2.1.jar org.gradle.launcher.daemon.bootstrap.GradleDaemon 3.2.1
```
i.e. Gradle daemon.

### Your Environment
------------------------------------------------------------
Gradle 3.2.1
------------------------------------------------------------

Build time:   2016-11-22 15:19:54 UTC
Revision:     83b485b914fd4f335ad0e66af9d14aad458d2cc5

Groovy:       2.4.7
Ant:          Apache Ant(TM) version 1.9.6 compiled on June 29 2015
JVM:          1.8.0_112 (Oracle Corporation 25.112-b15)
OS:           Windows 8.1 6.3 amd64

**Comment by sethjackson** [comment]
I'm experiencing the same problem with files added to classpath of a custom Ant task.

```gradle
task exampleTask {
    doLast {
        ant.taskdef(resource: 'taskdef.xml') {
            classpath {
                fileset(dir: 'example', includes: '*.jar')
            }
        }

        ant.customTask(example: 'test')
    }
}
```

Once the task is run some of the files in the `classpath` directive are still locked by the daemon.

Currently I'm working around this by using `--no-daemon`.

## Your Environment
------------------------------------------------------------
Gradle 3.2.1
------------------------------------------------------------

Build time:   2016-11-22 15:19:54 UTC
Revision:     83b485b914fd4f335ad0e66af9d14aad458d2cc5

Groovy:       2.4.7
Ant:          Apache Ant(TM) version 1.9.6 compiled on June 29 2015
JVM:          1.8.0_111 (Oracle Corporation 25.111-b14)
OS:           Windows 10 10.0 amd64

**Comment by eric-trousset** [comment]
Hi, 

I had the same issue with Gradle 3.4.1 building a android project under Jenkins/Win10

all subsequent build were failling due to Access Denied error when trying to cleanup the workspace

My work around is to call **gradlw.bat --stop** after the build.

**Comment by Diederikjh** [comment]
Similar workaround is to disable the daemon by adding  `org.gradle.daemon=false` to the  `gradle.properties` file.

**Comment by akarnokd** [comment]
I've encountered the same issue with 4.2-RC2 after upgrading from 4.0.1 in [one of my projects](https://github.com/akarnokd/Reactive4JavaFlow).

If I run 

`./gradlew clean jmh -Pjmh="ShakespearePlaysScrabbleWithFolyamOpt"` 

the generated jar file gets locked by one of the daemons and the next time I try to execute the same command line, it fails with access denied.

**Comment by Mashnin** [comment]
For me, this was [Hugo plugin](https://github.com/JakeWharton/hugo), which locked the files. I used [File Leak Detector](http://file-leak-detector.kohsuke.org/) to figure this out.

**Comment by SPessall** [comment]
We also have this problem. In our case it is caused by the [gradle-swagger-plugin](https://github.com/gigaSproule/swagger-gradle-plugin).

We are using gradle 4.0.

**Comment by KidDaedalus** [comment]
In our case this issue turned out to be the fault of the aspectj compiler.
https://discuss.gradle.org/t/tracking-down-unreleased-file-handles/23571

**Comment by timler** [comment]
One of my team members had this problem while compiling an Android project and we successfully used the workaround to avoid using a daemon. We don't want to permanently stop using the daemon for the project, and because it happens on each build, it's not possible to keep calling stop or killing the daemon.

Details:
```
$ ./gradlew --version

------------------------------------------------------------
Gradle 4.3.1
------------------------------------------------------------

Build time:   2017-11-08 08:59:45 UTC
Revision:     e4f4804807ef7c2829da51877861ff06e07e006d

Groovy:       2.4.12
Ant:          Apache Ant(TM) version 1.9.6 compiled on June 29 2015
JVM:          1.8.0_66 (Oracle Corporation 25.66-b17)
OS:           Mac OS X 10.11 x86_64
```

```
$ ./gradlew build --stacktrace

FAILURE: Build failed with an exception.

* Where:
Build file '/Users/xxxx/Projects/yyyyy/zzzzz/app/build.gradle' line: 1

* What went wrong:
A problem occurred evaluating project ':app'.
> Failed to apply plugin [id 'com.android.application']
   > Unable to initialize first usage tracking spool file

* Try:
Run with --info or --debug option to get more log output.

* Exception is:
org.gradle.api.GradleScriptException: A problem occurred evaluating project ':app'.
        at org.gradle.groovy.scripts.internal.DefaultScriptRunnerFactory$ScriptRunnerImpl.run(DefaultScriptRunnerFactory.java:92)
        at org.gradle.configuration.DefaultScriptPluginFactory$ScriptPluginImpl$2.run(DefaultScriptPluginFactory.java:209)
        at org.gradle.configuration.ProjectScriptTarget.addConfiguration(ProjectScriptTarget.java:77)
        at org.gradle.configuration.DefaultScriptPluginFactory$ScriptPluginImpl.apply(DefaultScriptPluginFactory.java:214)
        at org.gradle.configuration.BuildOperationScriptPlugin$1.run(BuildOperationScriptPlugin.java:61)
        at org.gradle.internal.progress.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:336)
        at org.gradle.internal.progress.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:328)
        at org.gradle.internal.progress.DefaultBuildOperationExecutor.execute(DefaultBuildOperationExecutor.java:199)
        at org.gradle.internal.progress.DefaultBuildOperationExecutor.run(DefaultBuildOperationExecutor.java:110)
        at org.gradle.configuration.BuildOperationScriptPlugin.apply(BuildOperationScriptPlugin.java:58)
        at org.gradle.configuration.project.BuildScriptProcessor.execute(BuildScriptProcessor.java:41)
        at org.gradle.configuration.project.BuildScriptProcessor.execute(BuildScriptProcessor.java:26)
        at org.gradle.configuration.project.ConfigureActionsProjectEvaluator.evaluate(ConfigureActionsProjectEvaluator.java:34)
        at org.gradle.configuration.project.LifecycleProjectEvaluator.doConfigure(LifecycleProjectEvaluator.java:64)
        at org.gradle.configuration.project.LifecycleProjectEvaluator.access$100(LifecycleProjectEvaluator.java:34)
        at org.gradle.configuration.project.LifecycleProjectEvaluator$ConfigureProject.run(LifecycleProjectEvaluator.java:110)
        at org.gradle.internal.progress.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:336)
        at org.gradle.internal.progress.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:328)
        at org.gradle.internal.progress.DefaultBuildOperationExecutor.execute(DefaultBuildOperationExecutor.java:199)
        at org.gradle.internal.progress.DefaultBuildOperationExecutor.run(DefaultBuildOperationExecutor.java:110)
        at org.gradle.configuration.project.LifecycleProjectEvaluator.evaluate(LifecycleProjectEvaluator.java:50)
        at org.gradle.api.internal.project.DefaultProject.evaluate(DefaultProject.java:656)
        at org.gradle.api.internal.project.DefaultProject.evaluate(DefaultProject.java:135)
        at org.gradle.execution.TaskPathProjectEvaluator.configure(TaskPathProjectEvaluator.java:35)
        at org.gradle.execution.TaskPathProjectEvaluator.configureHierarchy(TaskPathProjectEvaluator.java:62)
        at org.gradle.configuration.DefaultBuildConfigurer.configure(DefaultBuildConfigurer.java:38)
        at org.gradle.initialization.DefaultGradleLauncher$ConfigureBuild.run(DefaultGradleLauncher.java:249)
        at org.gradle.internal.progress.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:336)
        at org.gradle.internal.progress.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:328)
        at org.gradle.internal.progress.DefaultBuildOperationExecutor.execute(DefaultBuildOperationExecutor.java:199)
        at org.gradle.internal.progress.DefaultBuildOperationExecutor.run(DefaultBuildOperationExecutor.java:110)
        at org.gradle.initialization.DefaultGradleLauncher.configureBuild(DefaultGradleLauncher.java:167)
        at org.gradle.initialization.DefaultGradleLauncher.doBuildStages(DefaultGradleLauncher.java:126)
        at org.gradle.initialization.DefaultGradleLauncher.executeTasks(DefaultGradleLauncher.java:109)
        at org.gradle.internal.invocation.GradleBuildController$1.call(GradleBuildController.java:78)
        at org.gradle.internal.invocation.GradleBuildController$1.call(GradleBuildController.java:75)
        at org.gradle.internal.work.DefaultWorkerLeaseService.withLocks(DefaultWorkerLeaseService.java:152)
        at org.gradle.internal.invocation.GradleBuildController.doBuild(GradleBuildController.java:100)
        at org.gradle.internal.invocation.GradleBuildController.run(GradleBuildController.java:75)
        at org.gradle.tooling.internal.provider.ExecuteBuildActionRunner.run(ExecuteBuildActionRunner.java:28)
        at org.gradle.launcher.exec.ChainingBuildActionRunner.run(ChainingBuildActionRunner.java:35)
        at org.gradle.tooling.internal.provider.ValidatingBuildActionRunner.run(ValidatingBuildActionRunner.java:32)
        at org.gradle.launcher.exec.RunAsBuildOperationBuildActionRunner$1.run(RunAsBuildOperationBuildActionRunner.java:43)
        at org.gradle.internal.progress.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:336)
        at org.gradle.internal.progress.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:328)
        at org.gradle.internal.progress.DefaultBuildOperationExecutor.execute(DefaultBuildOperationExecutor.java:199)
        at org.gradle.internal.progress.DefaultBuildOperationExecutor.run(DefaultBuildOperationExecutor.java:110)
        at org.gradle.launcher.exec.RunAsBuildOperationBuildActionRunner.run(RunAsBuildOperationBuildActionRunner.java:40)
        at org.gradle.tooling.internal.provider.SubscribableBuildActionRunner.run(SubscribableBuildActionRunner.java:51)
        at org.gradle.launcher.exec.InProcessBuildActionExecuter.execute(InProcessBuildActionExecuter.java:47)
        at org.gradle.launcher.exec.InProcessBuildActionExecuter.execute(InProcessBuildActionExecuter.java:30)
        at org.gradle.launcher.exec.BuildTreeScopeBuildActionExecuter.execute(BuildTreeScopeBuildActionExecuter.java:39)
        at org.gradle.launcher.exec.BuildTreeScopeBuildActionExecuter.execute(BuildTreeScopeBuildActionExecuter.java:25)
        at org.gradle.tooling.internal.provider.ContinuousBuildActionExecuter.execute(ContinuousBuildActionExecuter.java:80)
        at org.gradle.tooling.internal.provider.ContinuousBuildActionExecuter.execute(ContinuousBuildActionExecuter.java:53)
        at org.gradle.tooling.internal.provider.ServicesSetupBuildActionExecuter.execute(ServicesSetupBuildActionExecuter.java:57)
        at org.gradle.tooling.internal.provider.ServicesSetupBuildActionExecuter.execute(ServicesSetupBuildActionExecuter.java:32)
        at org.gradle.tooling.internal.provider.GradleThreadBuildActionExecuter.execute(GradleThreadBuildActionExecuter.java:36)
        at org.gradle.tooling.internal.provider.GradleThreadBuildActionExecuter.execute(GradleThreadBuildActionExecuter.java:25)
        at org.gradle.tooling.internal.provider.ParallelismConfigurationBuildActionExecuter.execute(ParallelismConfigurationBuildActionExecuter.java:43)
        at org.gradle.tooling.internal.provider.ParallelismConfigurationBuildActionExecuter.execute(ParallelismConfigurationBuildActionExecuter.java:29)
        at org.gradle.tooling.internal.provider.StartParamsValidatingActionExecuter.execute(StartParamsValidatingActionExecuter.java:64)
        at org.gradle.tooling.internal.provider.StartParamsValidatingActionExecuter.execute(StartParamsValidatingActionExecuter.java:29)
        at org.gradle.tooling.internal.provider.SessionFailureReportingActionExecuter.execute(SessionFailureReportingActionExecuter.java:59)
        at org.gradle.tooling.internal.provider.SessionFailureReportingActionExecuter.execute(SessionFailureReportingActionExecuter.java:44)
        at org.gradle.tooling.internal.provider.SetupLoggingActionExecuter.execute(SetupLoggingActionExecuter.java:45)
        at org.gradle.tooling.internal.provider.SetupLoggingActionExecuter.execute(SetupLoggingActionExecuter.java:30)
        at org.gradle.launcher.daemon.server.exec.ExecuteBuild.doBuild(ExecuteBuild.java:67)
        at org.gradle.launcher.daemon.server.exec.BuildCommandOnly.execute(BuildCommandOnly.java:36)
        at org.gradle.launcher.daemon.server.api.DaemonCommandExecution.proceed(DaemonCommandExecution.java:122)
        at org.gradle.launcher.daemon.server.exec.WatchForDisconnection.execute(WatchForDisconnection.java:37)
        at org.gradle.launcher.daemon.server.api.DaemonCommandExecution.proceed(DaemonCommandExecution.java:122)
        at org.gradle.launcher.daemon.server.exec.ResetDeprecationLogger.execute(ResetDeprecationLogger.java:26)
        at org.gradle.launcher.daemon.server.api.DaemonCommandExecution.proceed(DaemonCommandExecution.java:122)
        at org.gradle.launcher.daemon.server.exec.RequestStopIfSingleUsedDaemon.execute(RequestStopIfSingleUsedDaemon.java:34)
        at org.gradle.launcher.daemon.server.api.DaemonCommandExecution.proceed(DaemonCommandExecution.java:122)
        at org.gradle.launcher.daemon.server.exec.ForwardClientInput$2.call(ForwardClientInput.java:74)
        at org.gradle.launcher.daemon.server.exec.ForwardClientInput$2.call(ForwardClientInput.java:72)
        at org.gradle.util.Swapper.swap(Swapper.java:38)
        at org.gradle.launcher.daemon.server.exec.ForwardClientInput.execute(ForwardClientInput.java:72)
        at org.gradle.launcher.daemon.server.api.DaemonCommandExecution.proceed(DaemonCommandExecution.java:122)
        at org.gradle.launcher.daemon.server.exec.LogAndCheckHealth.execute(LogAndCheckHealth.java:55)
        at org.gradle.launcher.daemon.server.api.DaemonCommandExecution.proceed(DaemonCommandExecution.java:122)
        at org.gradle.launcher.daemon.server.exec.LogToClient.doBuild(LogToClient.java:62)
        at org.gradle.launcher.daemon.server.exec.BuildCommandOnly.execute(BuildCommandOnly.java:36)
        at org.gradle.launcher.daemon.server.api.DaemonCommandExecution.proceed(DaemonCommandExecution.java:122)
        at org.gradle.launcher.daemon.server.exec.EstablishBuildEnvironment.doBuild(EstablishBuildEnvironment.java:82)
        at org.gradle.launcher.daemon.server.exec.BuildCommandOnly.execute(BuildCommandOnly.java:36)
        at org.gradle.launcher.daemon.server.api.DaemonCommandExecution.proceed(DaemonCommandExecution.java:122)
        at org.gradle.launcher.daemon.server.exec.StartBuildOrRespondWithBusy$1.run(StartBuildOrRespondWithBusy.java:50)
        at org.gradle.launcher.daemon.server.DaemonStateCoordinator$1.run(DaemonStateCoordinator.java:295)
        at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:63)
        at org.gradle.internal.concurrent.ManagedExecutorImpl$1.run(ManagedExecutorImpl.java:46)
        at org.gradle.internal.concurrent.ThreadFactoryImpl$ManagedThreadRunnable.run(ThreadFactoryImpl.java:55)
Caused by: org.gradle.api.internal.plugins.PluginApplicationException: Failed to apply plugin [id 'com.android.application']
        at org.gradle.api.internal.plugins.DefaultPluginManager.doApply(DefaultPluginManager.java:150)
        at org.gradle.api.internal.plugins.DefaultPluginManager.apply(DefaultPluginManager.java:125)
        at org.gradle.api.internal.plugins.DefaultObjectConfigurationAction.applyType(DefaultObjectConfigurationAction.java:120)
        at org.gradle.api.internal.plugins.DefaultObjectConfigurationAction.access$200(DefaultObjectConfigurationAction.java:38)
        at org.gradle.api.internal.plugins.DefaultObjectConfigurationAction$3.run(DefaultObjectConfigurationAction.java:86)
        at org.gradle.api.internal.plugins.DefaultObjectConfigurationAction.execute(DefaultObjectConfigurationAction.java:143)
        at org.gradle.api.internal.project.AbstractPluginAware.apply(AbstractPluginAware.java:46)
        at org.gradle.api.internal.project.ProjectScript.apply(ProjectScript.java:34)
        at org.gradle.api.Script$apply.callCurrent(Unknown Source)
        at build_3v35rx8rqdut34cnkis9634x4.run(/Users/xxxx/Projects/yyyyy/zzzzz/app/build.gradle:1)
        at org.gradle.groovy.scripts.internal.DefaultScriptRunnerFactory$ScriptRunnerImpl.run(DefaultScriptRunnerFactory.java:90)
        ... 93 more
Caused by: java.lang.RuntimeException: Unable to initialize first usage tracking spool file
        at com.android.tools.analytics.JournalingUsageTracker.<init>(JournalingUsageTracker.java:80)
        at com.android.tools.analytics.UsageTracker.initialize(UsageTracker.java:179)
        at com.android.builder.profile.ProcessProfileWriterFactory.initializeAnalytics(ProcessProfileWriterFactory.java:148)
        at com.android.builder.profile.ProcessProfileWriterFactory.get(ProcessProfileWriterFactory.java:161)
        at com.android.builder.profile.ProcessProfileWriterFactory.initialize(ProcessProfileWriterFactory.java:81)
        at com.android.build.gradle.internal.profile.ProfilerInitializer.init(ProfilerInitializer.java:68)
        at com.android.build.gradle.BasePlugin.apply(BasePlugin.java:257)
        at com.android.build.gradle.AppPlugin.apply(AppPlugin.java:120)
        at com.android.build.gradle.AppPlugin.apply(AppPlugin.java:47)
        at org.gradle.api.internal.plugins.ImperativeOnlyPluginTarget.applyImperative(ImperativeOnlyPluginTarget.java:42)
        at org.gradle.api.internal.plugins.RuleBasedPluginTarget.applyImperative(RuleBasedPluginTarget.java:50)
        at org.gradle.api.internal.plugins.DefaultPluginManager.addPlugin(DefaultPluginManager.java:165)
        at org.gradle.api.internal.plugins.DefaultPluginManager.access$200(DefaultPluginManager.java:47)
        at org.gradle.api.internal.plugins.DefaultPluginManager$AddPluginBuildOperation.run(DefaultPluginManager.java:252)
        at org.gradle.internal.progress.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:336)
        at org.gradle.internal.progress.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:328)
        at org.gradle.internal.progress.DefaultBuildOperationExecutor.execute(DefaultBuildOperationExecutor.java:199)
        at org.gradle.internal.progress.DefaultBuildOperationExecutor.run(DefaultBuildOperationExecutor.java:110)
        at org.gradle.api.internal.plugins.DefaultPluginManager.doApply(DefaultPluginManager.java:144)
        ... 103 more
Caused by: java.io.IOException: Unable to lock usage tracking spool file
        at com.android.tools.analytics.JournalingUsageTracker.newTrackFile(JournalingUsageTracker.java:98)
        at com.android.tools.analytics.JournalingUsageTracker.<init>(JournalingUsageTracker.java:78)
        ... 121 more
Caused by: java.nio.channels.OverlappingFileLockException
        at com.android.tools.analytics.JournalingUsageTracker.newTrackFile(JournalingUsageTracker.java:95)
        ... 122 more


* Get more help at https://help.gradle.org
```

**Comment by Mashnin** [comment]
@timler
What plugins do you use? I bet it's something AspectJ-related causing problems.

**Comment by timler** [comment]
@Mashnin it was my colleague's computer, but I doubt it's an aspectJ plugin because we have a very simple Android project and a vanilla Android Studio install

**Comment by Diab1o** [comment]
Getting the same problem with jmh plugin

**Comment by Argelbargel** [comment]
I have the same problem when using the checkstyle-plugin when using custom-checks in jar that is build by a subproject. First run (or any run with --no-daemon) works like a charm. But if the second run tries to e.g. clean before a build the build fails because the jar containing the custom-checks cannot be removed.

**Comment by myalcin81** [comment]
I have to same problem with @Argelbargel

**Comment by DaveXCS** [comment]
I have the same problem on Windows10 (with linux it seems to work).
After i run my build one time i have to use "gradle daemon --stop" to close the file handle.
In my case it will stop on a simple copy task:
`task copyZipInstall(type: Copy) {
    from zipUpdate
    into 'install/MC'
    doLast{
        println 'mmr update zip copied to install/MC'
    }
}`
Even if i try to copy the file manually i get access denied until i stop the daemon.
With the file-leak-detector i will see the problem as well:

`1 descriptors are open
#1 C:\Users\DP1016\.gradle\wrapper\dists\gradle-5.0-milestone-1-all\6dyq36h9cpyjh4c7hz2nzqfq6\gradle-5.0-milestone-1\lib\ant-1.9.13.jar by thread:Execution worker for ':' on Mon Oct 08 13:53:08 CEST 2018
	at java.util.zip.ZipFile.<init>(ZipFile.java:156)
	at java.util.jar.JarFile.<init>(JarFile.java:166)
	at java.util.jar.JarFile.<init>(JarFile.java:103)
	at sun.net.www.protocol.jar.URLJarFile.<init>(URLJarFile.java:93)
	at sun.net.www.protocol.jar.URLJarFile.getJarFile(URLJarFile.java:69)
	at sun.net.www.protocol.jar.JarFileFactory.get(JarFileFactory.java:109)
	at sun.net.www.protocol.jar.JarURLConnection.connect(JarURLConnection.java:122)
	at org.apache.tools.ant.taskdefs.Antlib.createAntlib(Antlib.java:69)
	at org.apache.tools.ant.taskdefs.Definer.loadAntlib(Definer.java:426)
	at org.apache.tools.ant.taskdefs.Definer.execute(Definer.java:278)
	at org.apache.tools.ant.ComponentHelper.checkNamespace(ComponentHelper.java:875)
	at org.apache.tools.ant.ComponentHelper.getDefinition(ComponentHelper.java:308)
	at org.apache.tools.ant.ComponentHelper.createComponent(ComponentHelper.java:285)
	at org.apache.tools.ant.ComponentHelper.createComponent(ComponentHelper.java:264)
	at org.apache.tools.ant.UnknownElement.makeObject(UnknownElement.java:429)
	at org.apache.tools.ant.UnknownElement.maybeConfigure(UnknownElement.java:164)
	at groovy.util.AntBuilder.performTask(AntBuilder.java:333)
	at groovy.util.AntBuilder.nodeCompleted(AntBuilder.java:283)
	at org.gradle.api.internal.project.ant.BasicAntBuilder.nodeCompleted(BasicAntBuilder.java:78)
	at groovy.util.BuilderSupport.doInvokeMethod(BuilderSupport.java:154)
	at groovy.util.AntBuilder.doInvokeMethod(AntBuilder.java:218)
	at org.gradle.api.internal.project.ant.BasicAntBuilder.doInvokeMethod(BasicAntBuilder.java:103)
	at groovy.util.BuilderSupport.invokeMethod(BuilderSupport.java:67)
	at org.codehaus.groovy.runtime.callsite.PogoMetaClassSite.call(PogoMetaClassSite.java:47)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:127)
	at build_6x7y05ui7ynrjweshv2m2p2wt$_run_closure6$_closure14.doCall(D:\PROGRAMMING\MasterController\web\build.gradle:75)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.codehaus.groovy.reflection.CachedMethod.invoke(CachedMethod.java:104)
	at groovy.lang.MetaMethod.doMethodInvoke(MetaMethod.java:326)
	at org.codehaus.groovy.runtime.metaclass.ClosureMetaClass.invokeMethod(ClosureMetaClass.java:264)
	at groovy.lang.MetaClassImpl.invokeMethod(MetaClassImpl.java:1041)
	at groovy.lang.Closure.call(Closure.java:411)
	at groovy.lang.Closure.call(Closure.java:427)`

**Comment by wszeboreq** [comment]
I'm experiencing the problem under Windows 7 with Gradle daemon locking the project directory of the project whose build caused the daemon to start. The only plugin I'm using is the 'maven-publish' plugin. I'm using Gradle 4.10.3. The issue causes Jenkins to be not able to clean the project workspace for the builds executed after the build which started the daemon.

The handle to the project dir is visible in ProcessExplorer under the Gradle daemon process. Manually closing the handle unlocks the project dir.

**Comment by grv87** [comment]
On JMH plugin, the problem may be a little different.

The `jmh` task, of type `JMHTask`, uses Worker API with provided classpath.
Classpath includes jar in `build` directory generated by separate task, `JmhBytecodeGeneratorTask`.
When daemon is used then there is another Gradle Worker daemon process for `JMHTask` that locks classpath files including the jar.
Generated jar has the same filename each time.
So, next time `JmhBytecodeGeneratorTask` is run it cannot rewrite a jar.
The same with `clean`.

See discussion at melix/jmh-gradle-plugin#145.

For me, it makes perfect sense that running application locks its binary files. (Strange, Linux doesn't do this??). This is why I say that the problem may be different.

1. I would like to ask Gradle team, whether this is a correct approach to use Worker API.
* Is it supposed to work with dynamic classpath?
* Maybe dynamic jars should have different filenames each time?
2. If this is a correct approach, then this is a bug. (Should I open another issue?)
Worker daemon should be stopped when its classpath is about to be regenerated.
3. Note that warm Gradle Worker daemon useful as long as sources of generated jar are not changing.
I'm not sure that it's worth to keep such daemon working.

**Comment by lacasseio** [comment]
@wszeboreq Can you confirm the same behaviour is experienced for the same scenario on the latest Gradle 5.2.1?

**Comment by wszeboreq** [comment]
@lacasseio I can't upgrade the Gradle version used in production quickly (we plan upgrading, but not now, as this will require some time and care). Anyway, I will try performing some manual experiments and let here now in case I have some information.

The thing I tried is starting the 5.2.1 daemon on Windows 10 via building project with 'empty.gradle' empty build script. Once build is finished, the Process Explorer still shows the unclosed handle to the project dir, if I'm interpreting this correctly - I'm attaching the screenshot 
![unclosed_daemon_project_dir_handle](https://user-images.githubusercontent.com/38911260/52634554-cc272d80-2ec7-11e9-8941-dfb0a1b11115.PNG)
 - 'gradle.limitations' is the project dir. The build was started in Cygwin.

**Comment by CauchyAbel** [comment]
I am still having same issue with Gradle 4.7 on Win 10. Has to use @eric-trousset work around and stop the daemon...

**Comment by waynix** [comment]
I have the same problem with gradle while building for android, we are already trying the --no-daemon option, this helped a bit, but the signing-config.json is still locked after build. 
My gradle Version:
```
------------------------------------------------------------
Gradle 5.1.1
------------------------------------------------------------

Build time:   2019-01-10 23:05:02 UTC
Revision:     3c9abb645fb83932c44e8610642393ad62116807

Kotlin DSL:   1.1.1
Kotlin:       1.3.11
Groovy:       2.5.4
Ant:          Apache Ant(TM) version 1.9.13 compiled on July 10 2018
JVM:          1.8.0_152-release (JetBrains s.r.o 25.152-b01)
OS:           Windows 10 10.0 amd64
```

**Comment by fleiber** [comment]
> @wszeboreq Can you confirm the same behaviour is experienced for the same scenario on the latest Gradle 5.2.1?

I can confirm that the issue is still present with all the latest versions: a simple project with a JMH benchmark, with jmh-gradle-plugin 0.5.0-rc1, Gradle 5.5.1, Windows 10 and JDK11.

**Comment by DaveXCS** [comment]
I took for me a long time until i found the problems on our site. We had a custom task which writes a xml file. One of the resource handlers was not closed so the daemon locked the file even the build was finished (had to kill the deamon before starting a new build).
After i closed everything with try with resources i have no problem anymore...

**Comment by fleiber** [comment]
In my case, it's a simple project not opening any resources (just a Gradle build with java and jmh plugins). The culprit is the jar generated by jmh.

**Comment by grv87** [comment]
@lacasseio, could you comment on my [question #1](https://github.com/gradle/gradle/issues/937#issuecomment-461857636)?
So that we could move somewhere with JMH plugin.

**Comment by maxxx** [comment]
Confirm on Win10 / Gradle 5.5.1 / Android Studio 3.5 / Git 2.23.0. Sometimes after git operations (mostly branches switch) one file in android project is locked. No JMH plugin or such things.

**Comment by tobymurray** [comment]
Adding an anecdote, using the same Gradle wrapper (5.6) on both Windows 10 (git version 2.23.0) and Ubuntu 18.04.3 (git version 2.17.1), I've been able to reproduce this only on Windows.

Gradle invoking Git on Ubuntu works just fine (no file lock), on Windows I have to separate out the tasks and invoke them with `--no-daemon` if I want to work around the file lock.

**Comment by fleiber** [comment]
> Adding an anecdote, using the same Gradle wrapper (5.6) on both Windows 10 (git version 2.23.0) and Ubuntu 18.04.3 (git version 2.17.1), I've been able to reproduce this only on Windows.
> 
> Gradle invoking Git on Ubuntu works just fine (no file lock), on Windows I have to separate out the tasks and invoke them with `--no-daemon` if I want to work around the file lock.

It's not surprising in fact: on Linux, the file locking doesn't work the same way. If a file held by a process is deleted, it succeeds and the file looks like it's gone... but it's not really, and its ghost stays until the process dies or releases it.
It just means your Gradle daemons end up holding more and more ghost files, the issue is just less visible.

**Comment by VolodymyrFesenko** [comment]
This is still an issue on Gradle 6.7

**Comment by cheppsn** [comment]
Can confirm, happens every time. Multi-module project (10 modules. Keeps locking the same module-jar whenever a new `clean assemble` build is started:

```
09:48:27: Executing tasks 'clean assemble'...

> Task :application:clean
> Task :backend-api:clean FAILED

...

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':backend-api:clean'.
> java.io.IOException: Unable to delete directory 'P:\projects\foo\backend\api\build'
    Failed to delete some children. This might happen because a process has files open or has its working directory set in the target directory.
    - P:\projects\foo\backend\api\build\libs\backend-api.jar
    - P:\projects\foo\backend\api\build\libs
```

**Comment by mictaege** [comment]
> Can confirm, happens every time. Multi-module project (10 modules. Keeps locking the same module-jar whenever a new `clean assemble` build is started:
> 
> ```
> 09:48:27: Executing tasks 'clean assemble'...
> 
> > Task :application:clean
> > Task :backend-api:clean FAILED
> 
> ...
> 
> FAILURE: Build failed with an exception.
> 
> * What went wrong:
> Execution failed for task ':backend-api:clean'.
> > java.io.IOException: Unable to delete directory 'P:\projects\foo\backend\api\build'
>     Failed to delete some children. This might happen because a process has files open or has its working directory set in the target directory.
>     - P:\projects\foo\backend\api\build\libs\backend-api.jar
>     - P:\projects\foo\backend\api\build\libs
> ```

The same thing happens to me in my multi-module project.

`>.\gradlew clean compileAll`

```
Execution failed for task ':ologis-api:clean'.
> java.io.IOException: Unable to delete directory 'X:\ologis\ologis-api'
    Failed to delete some children. This might happen because a process has files open or has its working directory set in the target directory.
    - X:\ologis\ologis-api\libs\ologis-api-NEXT-SNAPSHOT.jar
    - X:\ologis\ologis-api\libs
```

And as I figured out with [Sysinternals Handle](https://docs.microsoft.com/en-us/sysinternals/downloads/handle) and Java's _jps_ tool it is the gradle deamon itself - not any third party plugin - who locks the module jar.

`>handle64.exe ologis-api` 

```
Nthandle v4.22 - Handle viewer
Copyright (C) 1997-2019 Mark Russinovich
Sysinternals - www.sysinternals.com

java.exe           pid: 14808  type: File          1D4C: X:\ologis\ologis-api\libs\ologis-api-NEXT-SNAPSHOT.jar
java.exe           pid: 14808  type: File          1D4C: X:\ologis\ologis-api\libs\ologis-api-NEXT-SNAPSHOT.jar
```
`>jps -l`

```
15872
22720
25312
14808 org.gradle.launcher.daemon.bootstrap.GradleDaemon
27804 sun.tools.jps.Jps
```
Conclusion: _Handle_ shows that the only process who is locking `ologis-api-NEXT-SNAPSHOT.jar` has the pid **14808** and _jps_ shows that this is the **GradleDaemon** itself.

However, there is something special about my gradle configuration because I'm using a RAM drive with the drive letter _X:_ and are redirecting the _buildDir_ in the _init.gradle_ file: 

```
gradle.projectsLoaded {
    rootProject.allprojects {
        buildDir = "X:/${rootProject.name}/${project.name}"
    }
}
```
So I made another attempt by removing the redirecting configuration from the _init.gradle_ file but I still got the same IOException.

**Comment by coxsim** [comment]
This frequently happens on our Bamboo build agent host.  The only workaround is to kill the Gradle deamons and re-run failed builds.  The files Gradle are holding onto are always Git .pack files: e.g.

`C:\bamboo\xml-data\build-dir\CICD-******147-JOB1\******\.git\objects\pack\pack-7c57ba91d4434b1bca8f75c64e4d62908c768be4.pack`

**Comment by cruzel89** [comment]
This issue still exists in the current Gradle v7.4 (on Windows 10 x64 with openjdk v11.0.12).

In my case jUnit tests are failing due to this, because the [@TempDir](https://junit.org/junit5/docs/snapshot/user-guide/#writing-tests-built-in-extensions-TempDirectory) annotated temporary folder which is used as temporary `gradle.user.home` cannot be automatically deleted after the jUnit test was executed:

```
java.io.IOException: Failed to delete temp directory C:\Users\****\AppData\Local\Temp\junit492549456655448575. The following paths could not be deleted (see suppressed exceptions for details):
gradlehome,
gradlehome\caches,
gradlehome\caches\7.4,
gradlehome\caches\7.4\executionHistory,
gradlehome\caches\7.4\executionHistory\executionHistory.lock,
gradlehome\caches\7.4\fileContent,
gradlehome\caches\7.4\fileContent\fileContent.lock,
gradlehome\caches\7.4\fileHashes,
gradlehome\caches\7.4\fileHashes\fileHashes.bin,
gradlehome\caches\7.4\fileHashes\fileHashes.lock,
gradlehome\caches\7.4\fileHashes\resourceHashesCache.bin,
gradlehome\caches\jars-9,
... (some project-internal jars removed)
gradlehome\caches\jars-9\jars-9.lock,
gradlehome\caches\journal-1,
gradlehome\caches\journal-1\file-access.bin,
gradlehome\caches\journal-1\journal-1.lock,
gradlehome\caches\modules-2,
gradlehome\caches\modules-2\metadata-2.97,
gradlehome\caches\modules-2\metadata-2.97\module-artifact.bin,
gradlehome\caches\modules-2\metadata-2.97\module-artifacts.bin,
gradlehome\caches\modules-2\metadata-2.97\module-metadata.bin,
gradlehome\caches\modules-2\metadata-2.97\resource-at-url.bin,
gradlehome\caches\modules-2\modules-2.lock
```

The workaround in Groovy I am using now is to create the temporary folder manually via
```groovy
testProjectDir = Files.createTempDirectory("junit")
```
and force-delete it after the jUnit tests with:
```groovy
FileUtils.forceDeleteOnExit(testProjectDir.toFile())
```

But I would really appreciate if this issue would get fixed, so that I can use the [@TempDir](https://junit.org/junit5/docs/snapshot/user-guide/#writing-tests-built-in-extensions-TempDirectory) annotated folder functionality of jUnit 5.

@jjohannes Is it possible to reopen this issue or should I create a new issue instead?

**Comment by jjohannes** [comment]
@cruzel89 from my experiences I don't think there is still a general issue with the daemon (anymore). I would advice to open a new issue **with a reproducer** describing your exact setup if you really think this is an issue in Gradle that needs fixing.

What you describe sounds like you are running Gradle in a test. Are you testing a Gradle plugin/your own Gradle build setup?
In that case, what you see might be expected. The daemon stays alive and keeps files that are Gradle caches open on purpose. Possibly solution: run the Gradle-under-test using `--no-daemon`. Or you just do not put the gradle home in a temp folder. Usually you do not need that. It's certain not a "usual" setup in plugin/build testing.

If you really need to do such things you may check what Gradle's own test infrastructure does to test Gradle itself. You could have a look at the places that use [DaemonLogsAnalyzer](https://github.com/gradle/gradle/blob/b91bcc5f464cec06229df4420c9ae2ac79865fe9/subprojects/internal-integ-testing/src/main/groovy/org/gradle/integtests/fixtures/daemon/DaemonLogsAnalyzer.groovy) in Gradle's integration tests.

Note: I am not part of the Gradle team anymore. So there are only my two cents.

**Comment by cruzel89** [comment]
@jjohannes Thanks for the feedback!

Yes, I am testing a Gradle plugin via jUnit tests using the [TestKit library](https://docs.gradle.org/current/userguide/test_kit.html).
Unfortunately the `GradleRunner` of TestKit neither seems to run without Daemon when using `.withDebug(true)` (as proposed [here](https://discuss.gradle.org/t/testkit-how-to-turn-off-daemon/17843/2)) nor when applying `.withArguments("-Dorg.gradle.daemon=false")`. In both cases jUnit still fails to automatically delete the temp directory with an `java.io.IOException` after the test cases have been executed. Is there a different way of using `GradleRunner` without Daemon?

Btw: I just found a very similar issue described in #12535, which is still open.

**Comment by jjohannes** [comment]
@cruzel89 Yes #12535 is about the issue you are facing. I would suggest to comment/upvote on that issue.

**Comment by christhomas** [comment]
Does the `--project-cache-dir` option solve this problem by allowing multiple projects to share the .gradle directory for package artifacts but do all the project locking in a separate directory where they won't clash with each other?
